pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -38,18 +38,110 @@ var (
 )
 
 // A stateful listener builder
+// Support the below intentions
+// 1. Use separate inbound capture listener(:15006) and outbound capture listener(:15001)
+// 2. The above listeners use bind_to_port sub listeners or filter chains.
 type ListenerBuilder struct {
 	node                   *model.Proxy
 	gatewayListeners       []*xdsapi.Listener
 	inboundListeners       []*xdsapi.Listener
 	outboundListeners      []*xdsapi.Listener
 	virtualListener        *xdsapi.Listener
 	virtualInboundListener *xdsapi.Listener
+	useInboundFilterChain  bool
+}
+
+func insertOriginalListenerName(chain *listener.FilterChain, listenerName string) {
+	if chain.Metadata == nil {
+		chain.Metadata = &core.Metadata{
+			FilterMetadata: map[string]*types.Struct{},
+		}
+	}
+	if chain.Metadata.FilterMetadata[PilotMetaKey] == nil {
+		chain.Metadata.FilterMetadata[PilotMetaKey] = &types.Struct{
+			Fields: map[string]*types.Value{},
+		}
+	}
+	chain.Metadata.FilterMetadata[PilotMetaKey].Fields["original_listener_name"] =
+		&types.Value{Kind: &types.Value_StringValue{StringValue: listenerName}}
+}
+
+// Setup the filter chain match so that the match should work under both
+// - bind_to_port == false listener
+// - virtual inbound listener
+func amendFilterChainMatchFromInboundListener(chain *listener.FilterChain, l *xdsapi.Listener, needTLS bool) (*listener.FilterChain, bool) {
+	if chain.FilterChainMatch == nil {
+		chain.FilterChainMatch = &listener.FilterChainMatch{}
+	}
+	listenerAddress := l.Address
+	if sockAddr := listenerAddress.GetSocketAddress(); sockAddr != nil {
+		chain.FilterChainMatch.DestinationPort = &types.UInt32Value{Value: sockAddr.GetPortValue()}
+		if cidr := util.ConvertAddressToCidr(sockAddr.GetAddress()); cidr != nil {
+			if chain.FilterChainMatch.PrefixRanges != nil && len(chain.FilterChainMatch.PrefixRanges) != 1 {
+				log.Debugf("Intercepted inbound listener %s have neither 0 or 1 prefix ranges. Actual:  %d",
+					l.Name, len(chain.FilterChainMatch.PrefixRanges))
+			}
+			chain.FilterChainMatch.PrefixRanges = []*core.CidrRange{util.ConvertAddressToCidr(sockAddr.GetAddress())}
+		}
+		insertOriginalListenerName(chain, l.Name)
+	}
+	for _, filter := range l.ListenerFilters {
+		if needTLS = needTLS || filter.Name == xdsutil.TlsInspector; needTLS {
```

Comment:
what is `needTLS = needTLS`? seems like it is trying to do something but does nothing?

Reply:
I would interpret as needTLS = ( needTLS || filter.Name == xdsutil.TlsInspector) Let me see how to make it less confusing...

Reply:
1. ||= operator doesn't exist in golang
2. `b = b` is evaluated as bool so it doesn't confuse the compiler